### PR TITLE
Overshoot compensation

### PR DESCRIPTION
**Pull request by kev300**
My take on https://github.com/syssi/esphome-soyosource-gtn-virtual-meter/issues/73 😄 

powerc2

This graph shows the overshoot compensation in action with an update intervall of 0.5s. The powermeter provides new power readings a bit random between 0.7s and 2s (mostly 1s). The graph is for exactly that powermeter sensor.

It's not perfectly flat because the other inverter makes it spiky already and the time resolution of the graph is pretty hight. 😛 

With slow update interval of 3s (and no other inverter): https://github.com/syssi/esphome-soyosource-gtn-virtual-meter/issues/73#issuecomment-1465274313
Without compensation: https://github.com/syssi/esphome-soyosource-gtn-virtual-meter/issues/73#issuecomment-1460813033

I've added a new setting: max_power_sensor_latency_ms, which is 4000 (4s) by default.
That value also works for your defaults with 3s update_intervall. For higher update_intervalls this feature basically disables itself because of this value.
You can also set it to 0 to make sure it's disabled.

I did not add examples yet, because I don't know if you like it, yet. 😉 
Please review & test.

### CODE REVIEW

**Comment by syssi** [comment]
Could you rebase the PR?

File: components/soyosource_virtual_meter/soyosource_virtual_meter.cpp [hunk 1]
```diff
@@ -112,11 +112,11 @@ int16_t SoyosourceVirtualMeter::calculate_power_demand_negative_measurements_(in
     return this->max_power_demand_;
   }
 
-  if (power_demand > this->min_power_demand_) {
-    return power_demand;
+  if (power_demand < this->min_power_demand_) {
+    return this->min_power_demand_;
```
Comment (syssi) [inline components/soyosource_virtual_meter/soyosource_virtual_meter.cpp]: Could you move this fix into a separate PR?

Comment (kev300) [reply components/soyosource_virtual_meter/soyosource_virtual_meter.cpp]: ok: https://github.com/kev300/esphome-soyosource-gtn-virtual-meter/tree/min_power_demand_fix

File: components/soyosource_virtual_meter/soyosource_virtual_meter.cpp [hunk 2]
```diff
@@ -20,6 +20,11 @@ void SoyosourceVirtualMeter::setup() {
              this->last_power_demand_);
 
     this->last_power_demand_received_ = millis();
+
+    //if update intervall is set to never, trigger the update() method from here
+    if (this->get_update_interval() == -1) {
```
Comment (syssi) [inline components/soyosource_virtual_meter/soyosource_virtual_meter.cpp:25]:
```suggestion
    if (this->get_update_interval() == SCHEDULER_DONT_RUN) {
```

See esphome/core/component.h

```
static const uint32_t SCHEDULER_DONT_RUN = 4294967295UL;
```

Comment (kev300) [reply components/soyosource_virtual_meter/soyosource_virtual_meter.cpp]: yeah new c++ stuff for me. 😁 (did not code in c++ before this project.

File: components/soyosource_virtual_meter/soyosource_virtual_meter.cpp [hunk 3]
```diff
@@ -20,6 +20,11 @@ void SoyosourceVirtualMeter::setup() {
              this->last_power_demand_);
 
     this->last_power_demand_received_ = millis();
+
+    //if update intervall is set to never, trigger the update() method from here
```
Comment (syssi) [inline components/soyosource_virtual_meter/soyosource_virtual_meter.cpp:24]:
```suggestion
    // If the update_interval is set to never, call the update() method on every power sensor update
```

**Comment by syssi** [comment]
Could you apply `clang-format -i *.cpp *.h` on the touched files and rebase another time? :-)

This is another code style issue which must be fixed manually:

https://github.com/syssi/esphome-soyosource-gtn-virtual-meter/actions/runs/4436969951/jobs/7786087270

Just remove the "_" at the end of the variable name. It's a local variable and no class property.

File: components/soyosource_virtual_meter/soyosource_virtual_meter.cpp (hunk 3 above)
Comment (kev300) [reply components/soyosource_virtual_meter/soyosource_virtual_meter.cpp]: good suggestion, can't commit it right now. I guess you need to update it.

**Comment by syssi** [comment]
In general I'm happy about your approach because the virtualmeter will be responsible for the timing in future. I will need some time for testing and we should care about the following questions too:

1. Is this a breaking change?
2. Should we keep the old behaviour too?
3. What about people with slow power meter sensor updates? `update_interval: never` won't work here because a RS485 power demand message is required every few seconds.
4. What's the magic demand delta? It's something like impedance?

File: components/soyosource_virtual_meter/soyosource_virtual_meter.cpp [hunk 4]
```diff
@@ -20,6 +20,11 @@ void SoyosourceVirtualMeter::setup() {
              this->last_power_demand_);
 
     this->last_power_demand_received_ = millis();
+
+    // if update intervall is set to never, trigger the update() method from here
```
Comment (syssi) [inline components/soyosource_virtual_meter/soyosource_virtual_meter.cpp:24]:
```suggestion
    // If the update_interval is set to never, call the update() method on every power sensor update
```

**Comment by kev300** [comment]
> In general I'm happy about your approach because the virtualmeter will be responsible for the timing in future. I will need some time for testing and we should care about the following questions too:
> 
>     1. Is this a breaking change?

No, it's compatible with the old config, where update_intervall is set. 

>     2. Should we keep the old behaviour too?
Yes, it's still there. The virtualmeter will only trigger if update_intervall is not set or set to never.
If update_intervall is set to some value, the old behavior is in place.

>     3. What about people with slow power meter sensor updates? `update_interval: never` won't work here because a RS485 power demand message is required every few seconds.
They should keep their update_interval to 3s. See above.

>     4. What's the magic demand delta? It's something like impedance?
I simply did not found a better name for my formula with multiplication by 0.8.
I'm decreasing the power_demand_delta_, so that the compensation can stop compensating a bit earlier than it would with just power_demand_delta_. 
The theory is that at some point the consumption measurement should reach the same delta as the previosly set demand delta. Then we know, that the change in demand actually reached the measurement meter.

But in reality there is other stuff turning on and off at the same time, which can drift the measurement away and we won't reach the same delta ever. With the reduction, the compensation can stop, when the change is close enough. In addition to that the timeout makes sure, that at some point the compensation always stops. 

Why 80%? I had good results in my tests with it. Maybe another value will work better for you. I was thinking to make it configurable, but it's hard to decide which value to set there.

**Comment by syssi** [comment]
Could you explain why did you set the `update_interval` to `never` at all? The update method is used to send rs485 power demands periodically only. I would like to stick to an update interval between 0.5 and 5s here because it's required for the `manual` mode.

**Comment by kev300** [comment]
Because I wanted to match the sensor reading with the demand setting. So each time I get a new value, a new demand can be calculated and set. And since my sensor is read each second, its often enough (but has inconsistancy issues as mentioned).

The manual mode still works then, because the new value triggers everything and the update method decides to keep demand at 0 in manual mode. So it's not required.

But you can stick to a set intervall, because it's more reliable. But I think it is a good idea to have that option. 😄

**Comment by syssi** [comment]
Good point. So we shouldn't enforce any restrictions/limits here.

**Comment by kev300** [comment]
How is it going? Did you have a chance to test this?

My "system" is finally operational (meaning connected to solar an working), so I found and fix some bugs. 😬

**Comment by syssi** [comment]
Good job! I'm very happy about the improved naming!

**Comment by kev300** [comment]
Hey, don't hurry with the testing. I think i have a simpler and better working impl again. 😬 
But I'm currently testing it myself.

This aproach here still had a few overshoots from time to time, because it did not consider up ramping consumption changes.
